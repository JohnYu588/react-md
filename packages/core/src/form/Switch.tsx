import { cnb } from "cnbuilder";
import {
  forwardRef,
  type CSSProperties,
  type HTMLAttributes,
  type InputHTMLAttributes,
  type ReactNode,
} from "react";
import { type PropsWithRef } from "../types.js";
import { useEnsuredId } from "../useEnsuredId.js";
import { FormMessageContainer } from "./FormMessageContainer.js";
import { type InputToggleLabelProps } from "./InputToggle.js";
import { Label } from "./Label.js";
import { SwitchTrack } from "./SwitchTrack.js";
import { switchStyles } from "./switchStyles.js";
import {
  type FormComponentStates,
  type FormMessageContainerExtension,
} from "./types.js";

/**
 * @since 6.0.0 Added `containerProps` and support for the
 * `FormMessage` behavior.
 */
export interface SwitchProps
  extends InputHTMLAttributes<HTMLInputElement>,
    InputToggleLabelProps,
    FormMessageContainerExtension,
    FormComponentStates {
  containerProps?: PropsWithRef<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;
  trackProps?: PropsWithRef<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;
  trackStyle?: CSSProperties;
  trackClassName?: string;
  ballAddon?: ReactNode;
  ballProps?: PropsWithRef<HTMLAttributes<HTMLSpanElement>, HTMLSpanElement>;
  ballStyle?: CSSProperties;
  ballClassName?: string;
}

/**
 * @example Simple Example
 * ```tsx
 * import { Form, Switch } from "@react-md/core";
 * import type { ReactElement } from "react";
 * import { useState } from "react";
 *
 * function Example(): ReactElement {
 *   const [checked, setChecked] = useState(false);
 *
 *   return (
 *     <Form>
 *       <Switch
 *         label="Label"
 *         name="enabled"
 *         checked={checked}
 *         onChange={(event) => setChecked(event.currentTarget.checked)}
 *       />
 *     </Form>
 *   );
 * }
 * ```
 *
 * @since 6.0.0 Added support for `FormMessage` behavior.
 */
export const Switch = forwardRef<HTMLInputElement, SwitchProps>(
  function Switch(props, ref) {
    const {
      id: propId,
      label,
      labelProps,
      style,
      className,
      containerProps,
      ballProps,
      ballStyle,
      ballClassName,
      ballAddon,
      trackProps,
      trackStyle,
      trackClassName,
      messageProps,
      messageContainerProps,
      disableLabelGap = false,
      error = false,
      active = false,
      stacked = false,
      iconAfter = false,
      disabled = false,
      readOnly = false,
      ...remaining
    } = props;
    const id = useEnsuredId(propId, "switch");

    return (
      <FormMessageContainer
        {...messageContainerProps}
        messageProps={messageProps}
      >
        <Label
          style={style}
          {...labelProps}
          className={cnb(className, labelProps?.className)}
          gap={!disableLabelGap}
          stacked={stacked}
          reversed={!iconAfter}
          active={active}
          error={error}
          disabled={disabled}
        >
          {label}
          <div
            {...containerProps}
            className={switchStyles({
              clickable: !disabled && !readOnly,
              className: containerProps?.className,
              currentColor: active || error,
            })}
          >
            <SwitchTrack
              style={trackStyle}
              {...trackProps}
              className={cnb(trackClassName, trackProps?.className)}
              disabled={disabled}
              ballAddon={ballAddon}
              ballProps={ballProps}
              ballStyle={ballStyle}
              ballClassName={ballClassName}
            >
              <input
                {...remaining}
                id={id}
                ref={ref}
                role="switch"
                type="checkbox"
                className={cnb(
                  "rmd-switch__input rmd-hidden-input",
                  disabled && "rmd-hidden-input--disabled"
                )}
                disabled={disabled}
              />
            </SwitchTrack>
          </div>
        </Label>
      </FormMessageContainer>
    );
  }
);
