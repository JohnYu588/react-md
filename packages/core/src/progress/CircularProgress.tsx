import { cnb } from "cnbuilder";
import {
  type CSSProperties,
  type HTMLAttributes,
  forwardRef,
  useMemo,
} from "react";

import { cssUtils } from "../cssUtils.js";
import { type LabelRequiredForA11y } from "../types.js";
import { useEnsuredId } from "../useEnsuredId.js";
import { bem } from "../utils/bem.js";
import { getPercentage } from "../utils/getPercentage.js";
import { type ProgressProps } from "./types.js";

/**
 * @since 6.0.0 Added the `disableShrink` prop.
 * @since 6.0.0 Renamed `small` to `dense`.
 * @since 6.0.0 Renamed `centered` to `disableCentered`.
 * @since 6.0.0 Removed the `maxRotation` prop since the determinate state no
 * longer rotates while increasing value.
 */
export interface CircularProgressProps
  extends Omit<HTMLAttributes<HTMLSpanElement>, "id">,
    ProgressProps {
  /**
   * An optional style to apply to the svg within the circular progress. The
   * values of this style object will be merged with the current determinate
   * style (if it exists).
   */
  svgStyle?: CSSProperties;

  /**
   * An optional className to apply to the svg within the circular progress.
   */
  svgClassName?: string;

  /**
   * An optional style to apply to the circle within the circular progress. The
   * values of this style object will be merged with the current determinate
   * style (if it exists).
   */
  circleStyle?: CSSProperties;

  /**
   * An optional className to apply to the circle within the circular progress.
   */
  circleClassName?: string;

  /**
   * The radius for the circle. It is generally recommended to have the radius
   * be 1/2 of the viewbox and minus a few more pixels so that there is some
   * surrounding padding. You probably shouldn't really be changing this prop
   * though.
   *
   * @defaultValue `30`
   */
  radius?: number;

  /**
   * The center point for the circle. This should be half of the `viewBox` prop
   * 99% of the time and probably won't be changed.
   *
   * @defaultValue `33`
   */
  center?: number;

  /**
   * The viewbox for the child svg. I wouldn't recommend changing this value as
   * you will also need to update the `dashoffset` in both Sass and this prop to
   * get the animation to look nice again.
   *
   * @defaultValue `0 0 66 66`
   */
  viewBox?: string;

  /**
   * The `stroke-dashoffset` for the circle within the SVG. You probably won't
   * be changing this value that much as it should match the
   * `$rmd-progress-circle-dashoffset` Sass variable. This is really just used
   * to help to create the determinate progress animation.
   *
   * @defaultValue `187`
   */
  dashoffset?: number;

  /**
   * Boolean if the circular progress should be centered using left and right
   * margins.
   *
   * @defaultValue `false`
   * @since 6.0.0 Renamed from `centered`
   */
  disableCentered?: boolean;

  /**
   * Set to `true` to render as a smaller size.
   *
   * @defaultValue `false`
   * @since 2.3.0
   * @since 6.0.0 Renamed from `small`
   */
  dense?: boolean;

  /**
   * Set this to `true` to update the indeterminate behavior to only rotate
   * which will increase performance during CPU-intensive tasks or when many
   * loading spinners are displayed at once on the page.
   *
   * @defaultValue `false`
   * @since 6.0.0
   */
  disableShrink?: boolean;
}

const styles = bem("rmd-circular-progress");

/**
 * @example Indeterminate Example
 * ```tsx
 * import { CircularProgress } from "@react-md/core/progress/CircularProgress":
 * import { type ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return <CircularProgress aria-label="Loading" />;
 * }
 * ```
 *
 * @example Determinate Example
 * ```tsx
 * import { CircularProgress } from "@react-md/core/progress/CircularProgress":
 * import { useState, type ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   // a number from 0 - 100
 *   const [progress, setProgress] = useState(0);
 *
 *   return <CircularProgress aria-label="File upload" value={progress} />;
 * }
 * ```
 *
 * @since 6.0.0 Updated the determinate circular progress to no longer
 * rotate while increasing the value and require a label for accessibility.
 */
export const CircularProgress = forwardRef<
  HTMLSpanElement,
  LabelRequiredForA11y<CircularProgressProps>
>(function CircularProgress(props, ref) {
  const {
    id: propId,
    className,
    svgStyle,
    svgClassName,
    circleStyle: propCircleStyle,
    circleClassName,
    value,
    min = 0,
    max = 100,
    radius = 30,
    center = 33,
    viewBox = "0 0 66 66",
    theme = "primary",
    dense = false,
    dashoffset = 187,
    disableShrink = false,
    disableCentered = false,
    disableTransition = false,
    ...remaining
  } = props;

  const id = useEnsuredId(propId, "circular-progress");
  let progress: number | undefined;
  if (typeof value === "number") {
    progress = getPercentage({ min, max, value, validate: true });
  }

  const circleStyle = useMemo(() => {
    if (typeof progress !== "number") {
      return propCircleStyle;
    }

    return {
      ...propCircleStyle,
      strokeDashoffset: dashoffset - dashoffset * progress,
    };
  }, [progress, propCircleStyle, dashoffset]);

  const determinate = typeof progress === "number";
  const indeterminate = !determinate;
  return (
    <span
      {...remaining}
      id={id}
      ref={ref}
      role="progressbar"
      aria-valuemin={min}
      aria-valuemax={max}
      aria-valuenow={value}
      className={cnb(
        styles({ dense, centered: !disableCentered }),
        theme !== "current-color" && cssUtils({ textColor: theme }),
        className
      )}
    >
      <svg
        style={svgStyle}
        className={cnb(
          styles("svg", {
            determinate,
            indeterminate: indeterminate && !disableShrink,
            "rotate-only": indeterminate && disableShrink,
          }),
          svgClassName
        )}
        viewBox={viewBox}
      >
        <circle
          style={circleStyle}
          className={cnb(
            styles("circle", {
              animate: !disableTransition && determinate,
              determinate,
              indeterminate: indeterminate && !disableShrink,
              "rotate-only": indeterminate && disableShrink,
            }),
            circleClassName
          )}
          r={radius}
          cx={center}
          cy={center}
        />
      </svg>
    </span>
  );
});
