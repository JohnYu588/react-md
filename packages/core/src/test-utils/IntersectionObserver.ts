export class IntersectionObserverMock implements IntersectionObserver {
  root: Document | Element | null;
  rootMargin: string;
  thresholds: readonly number[];

  elements: Set<Element>;

  constructor(
    public callback: IntersectionObserverCallback,
    options: IntersectionObserverInit = {}
  ) {
    this.root = options.root || null;
    this.rootMargin = options.rootMargin || "";
    this.thresholds =
      typeof options.threshold === "number"
        ? [options.threshold]
        : options.threshold ?? [];

    this.elements = new Set();
  }

  observe(target: Element): void {
    this.elements.add(target);

    this.callback(this.takeRecords(), this);
  }

  unobserve(target: Element): void {
    this.elements.delete(target);
  }

  takeRecords(): IntersectionObserverEntry[] {
    return [...this.elements].map<IntersectionObserverEntry>((target) => ({
      time: Date.now(),
      target,
      boundingClientRect: target.getBoundingClientRect(),
      intersectionRatio: 0,
      intersectionRect: target.getBoundingClientRect(),
      isIntersecting: false,
      rootBounds:
        this.root && "getBoundingClientRect" in this.root
          ? this.root.getBoundingClientRect()
          : null,
    }));
  }

  disconnect(): void {
    this.elements.clear();
  }
}
