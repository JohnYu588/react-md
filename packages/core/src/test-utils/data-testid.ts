/// <reference types="react" />

declare module "react" {
  interface HTMLAttributes<T>
    extends React.AriaAttributes,
      React.DOMAttributes<T> {
    "data-testid"?: string | number;
  }
}
