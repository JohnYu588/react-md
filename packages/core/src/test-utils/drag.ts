import { fireEvent } from "@testing-library/dom";
import type { MouseEvent } from "react";
import { wait } from "../utils/wait.js";

interface XYCoords {
  x: number;
  y: number;
}

interface BaseDragOptions {
  to?: XYCoords | Element;
  delta?: XYCoords;
  steps?: number;
  duration?: number;
}

type DragOptions = BaseDragOptions &
  ({ to: XYCoords | Element; delta?: never } | { delta: XYCoords; to?: never });

const getElementClientCenter = (element: Element): XYCoords => {
  const { left, top, width, height } = element.getBoundingClientRect();
  return {
    x: left + width / 2,
    y: top + height / 2,
  };
};

const getCoords = (elementOrCoords: Element | XYCoords): XYCoords =>
  "x" in elementOrCoords && "y" in elementOrCoords
    ? elementOrCoords
    : getElementClientCenter(elementOrCoords);

/**
 * This is internal for now. No idea if it's actually worthwhile though since
 * you have to mock the `getBoundingClientRect()` to work.
 *
 * @internal
 * @see https://github.com/testing-library/testing-library-docs/blob/c20f8a8931e9f0a158fda9f01bbea8cb45ed1651/docs/example-drag.mdx
 */
export async function drag(
  element: Element,
  options: DragOptions
): Promise<void> {
  const { delta, to: inTo, steps = 20, duration = 500 } = options;
  const from = getElementClientCenter(element);
  const to = delta
    ? {
        x: from.x + delta.x,
        y: from.y + delta.y,
      }
    : getCoords(inTo);

  const step: XYCoords = {
    x: (to.x - from.x) / steps,
    y: (to.y - from.y) / steps,
  };

  const currentEvent = {
    clientX: from.x,
    clientY: from.y,
    button: 0,
    altKey: false,
    ctrlKey: false,
    metaKey: false,
    shiftKey: false,
  } satisfies Partial<MouseEvent>;

  fireEvent.mouseEnter(element, currentEvent);
  fireEvent.mouseOver(element, currentEvent);
  fireEvent.mouseMove(element, currentEvent);
  fireEvent.mouseDown(element, currentEvent);
  fireEvent.dragStart(element, currentEvent);

  const delay = duration / steps;
  for (let i = 0; i < steps; i++) {
    currentEvent.clientX += step.x;
    currentEvent.clientY += step.y;
    await wait(delay);
    fireEvent.mouseMove(element, currentEvent);
  }

  fireEvent.mouseUp(element, currentEvent);
  fireEvent.dragEnd(element, currentEvent);
}
