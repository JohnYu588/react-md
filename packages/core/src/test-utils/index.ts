export * from "@testing-library/react";
export * from "@testing-library/user-event";
export * from "./mocks/match-media.js";
export * from "./render.js";
export * from "./utils/resize-observer.js";
