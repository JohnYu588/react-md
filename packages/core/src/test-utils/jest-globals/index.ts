export * from "./match-media.js";
export * from "./resize-observer.js";
export * from "./timers.js";
