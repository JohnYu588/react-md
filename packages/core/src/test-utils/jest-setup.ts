import { beforeEach } from "@jest/globals";
import { INTERACTION_CONFIG } from "../interaction/config.js";

beforeEach(() => {
  // set the mode to `none` in tests since ripples require
  // `getBoundingClientRect()` to create correct CSS. You'll either see warnings
  // in the console around invalid css values or `NaN`.
  INTERACTION_CONFIG.mode = "none";
});
