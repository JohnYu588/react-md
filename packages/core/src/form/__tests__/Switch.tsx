import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { CircularProgress } from "../../progress/CircularProgress.js";
import { rmdRender, screen } from "test-utils";
import { Switch, type SwitchProps } from "../Switch.js";

const labelProps = {
  "data-testid": "label",
};

describe("Switch", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLInputElement>();
    const props = {
      ref,
      label: "Switch",
      labelProps,
    } as const;

    const { rerender } = rmdRender(<Switch {...props} />);

    const label = screen.getByTestId("label");
    const switchInput = screen.getByRole("switch", { name: "Switch" });
    expect(ref.current).toBeInstanceOf(HTMLInputElement);
    expect(ref.current).toBe(switchInput);
    expect(label).toMatchSnapshot();

    rerender(
      <Switch
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(label).toMatchSnapshot();

    rerender(
      <Switch
        {...props}
        ballStyle={{ background: "orange" }}
        ballClassName="ball-class-name"
        trackStyle={{ width: "10%" }}
        trackClassName="track-class-name"
        trackProps={{
          className: "track-props-class-name",
        }}
        containerProps={{
          style: { fontSize: "3rem" },
          className: "container-class-name",
        }}
      />
    );
    expect(label).toMatchSnapshot();

    rerender(<Switch {...props} id="custom-id" />);
    expect(label).toMatchSnapshot();
  });

  it("should allow the switch to gain the error and active colors by enabling the current-color class", () => {
    const { rerender } = rmdRender(
      <Switch label="Switch" active labelProps={labelProps} />
    );
    const label = screen.getByTestId("label");
    expect(label).toMatchSnapshot();

    rerender(<Switch label="Switch" error />);
    expect(label).toMatchSnapshot();
  });

  it("should merge any labelProps correctly", () => {
    rmdRender(
      <Switch
        label="Switch"
        style={{ color: "orange" }}
        labelProps={{
          ...labelProps,
          style: { color: "red" },
          className: "test-class-name",
        }}
        className="root-class-name"
      />
    );

    const label = screen.getByTestId("label");
    expect(label).toMatchSnapshot();
  });

  it("should render correctly when disabled", () => {
    rmdRender(<Switch label="Switch" labelProps={labelProps} disabled />);

    const label = screen.getByTestId("label");
    expect(label).toMatchSnapshot();
  });

  it("should support rendering the label before and after as well as stacked", () => {
    const { rerender } = rmdRender(
      <Switch label="Switch" labelProps={labelProps} iconAfter />
    );

    const label = screen.getByTestId("label");
    expect(label).toMatchSnapshot();

    rerender(<Switch label="Switch" labelProps={labelProps} stacked />);
    expect(label).toMatchSnapshot();

    rerender(
      <Switch label="Switch" labelProps={labelProps} stacked iconAfter />
    );
    expect(label).toMatchSnapshot();
  });

  it("should support rendering the form messages with the switch", () => {
    const props: SwitchProps = {
      label: "Label",
      messageProps: {
        children: "Help Text",
      },
      messageContainerProps: {
        "data-testid": "container",
      },
    };

    const { rerender } = rmdRender(<Switch {...props} />);

    const container = screen.getByTestId("container");
    expect(() => screen.getByText("Help Text")).not.toThrow();
    expect(container).toMatchSnapshot();

    rerender(
      <Switch
        {...props}
        error
        messageProps={{ error: true, children: "Error Text" }}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should support rendering a CircularProgress by using the ballAddon prop", () => {
    rmdRender(
      <Switch
        label="Switch"
        labelProps={labelProps}
        ballAddon={<CircularProgress aria-label="Loading" />}
      />
    );

    expect(() => screen.getByRole("progressbar")).not.toThrow();
    const label = screen.getByTestId("label");
    expect(label).toMatchSnapshot();
  });
});
