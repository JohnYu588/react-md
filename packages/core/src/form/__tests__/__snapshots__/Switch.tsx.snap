// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should allow the switch to gain the error and active colors by enabling the current-color class 1`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--active rmd-label--reversed"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch rmd-switch--current-color rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r1:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should allow the switch to gain the error and active colors by enabling the current-color class 2`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--error rmd-label--reversed rmd-error-color"
>
  Switch
  <div
    class="rmd-switch rmd-switch--current-color rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r1:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--reversed"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r0:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--reversed custom-class-name"
  data-testid="label"
  style="color: white;"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r0:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should apply the correct styling, HTMLAttributes, and allow a ref 3`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--reversed"
  data-testid="label"
  style=""
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable container-class-name"
    style="font-size: 3rem;"
  >
    <div
      class="rmd-switch__track track-class-name track-props-class-name"
      style="width: 10%;"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r0:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball ball-class-name"
        style="background: orange;"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should apply the correct styling, HTMLAttributes, and allow a ref 4`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--reversed"
  data-testid="label"
  style=""
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
    style=""
  >
    <div
      class="rmd-switch__track"
      style=""
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="custom-id"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
        style=""
      />
    </div>
  </div>
</label>
`;

exports[`Switch should merge any labelProps correctly 1`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--reversed root-class-name test-class-name"
  data-testid="label"
  style="color: red;"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r2:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should render correctly when disabled 1`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--disabled rmd-label--reversed rmd-text-disabled-color"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch"
  >
    <div
      class="rmd-switch__track rmd-switch__track--disabled"
    >
      <input
        class="rmd-switch__input rmd-hidden-input rmd-hidden-input--disabled"
        disabled=""
        id="switch-:r3:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should support rendering a CircularProgress by using the ballAddon prop 1`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--reversed"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r7:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      >
        <span
          aria-label="Loading"
          aria-valuemax="100"
          aria-valuemin="0"
          class="rmd-circular-progress rmd-circular-progress--centered rmd-primary-color"
          id="circular-progress-:r8:"
          role="progressbar"
        >
          <svg
            class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
            viewBox="0 0 66 66"
          >
            <circle
              class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
              cx="33"
              cy="33"
              r="30"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</label>
`;

exports[`Switch should support rendering the form messages with the switch 1`] = `
<div
  class="rmd-form-message-container"
  data-testid="container"
>
  <label
    class="rmd-label rmd-label--gap rmd-label--reversed"
  >
    Label
    <div
      class="rmd-switch rmd-switch--clickable"
    >
      <div
        class="rmd-switch__track"
      >
        <input
          class="rmd-switch__input rmd-hidden-input"
          id="switch-:r5:"
          role="switch"
          type="checkbox"
        />
        <span
          class="rmd-switch__ball"
        />
      </div>
    </div>
  </label>
  <div
    aria-live="polite"
    class="rmd-form-message rmd-form-message--outline rmd-text-secondary-color"
    id="form-message-:r6:"
  >
    <p
      class="rmd-form-message__message"
      id="form-message-:r6:-message"
    >
      Help Text
    </p>
  </div>
</div>
`;

exports[`Switch should support rendering the form messages with the switch 2`] = `
<div
  class="rmd-form-message-container"
  data-testid="container"
>
  <label
    class="rmd-label rmd-label--gap rmd-label--error rmd-label--reversed rmd-error-color"
  >
    Label
    <div
      class="rmd-switch rmd-switch--current-color rmd-switch--clickable"
    >
      <div
        class="rmd-switch__track"
      >
        <input
          class="rmd-switch__input rmd-hidden-input"
          id="switch-:r5:"
          role="switch"
          type="checkbox"
        />
        <span
          class="rmd-switch__ball"
        />
      </div>
    </div>
  </label>
  <div
    aria-live="polite"
    class="rmd-form-message rmd-form-message--outline rmd-error-color"
    id="form-message-:r6:"
  >
    <p
      class="rmd-form-message__message"
      id="form-message-:r6:-message"
    >
      Error Text
    </p>
  </div>
</div>
`;

exports[`Switch should support rendering the label before and after as well as stacked 1`] = `
<label
  class="rmd-label rmd-label--gap"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r4:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should support rendering the label before and after as well as stacked 2`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--stacked-reversed"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r4:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;

exports[`Switch should support rendering the label before and after as well as stacked 3`] = `
<label
  class="rmd-label rmd-label--gap rmd-label--stacked"
  data-testid="label"
>
  Switch
  <div
    class="rmd-switch rmd-switch--clickable"
  >
    <div
      class="rmd-switch__track"
    >
      <input
        class="rmd-switch__input rmd-hidden-input"
        id="switch-:r4:"
        role="switch"
        type="checkbox"
      />
      <span
        class="rmd-switch__ball"
      />
    </div>
  </div>
</label>
`;
